File: src/main/java/com/pauljoda/modularsystems/storage/container/ContainerStorageCore.java
package com.pauljoda.modularsystems.storage.container;

import com.pauljoda.modularsystems.storage.tiles.TileStorageCore;
import com.teambr.bookshelf.inventory.BaseContainer;
import cpw.mods.fml.common.registry.GameRegistry;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.entity.player.InventoryPlayer;
import net.minecraft.inventory.*;
import net.minecraft.item.ItemStack;
import net.minecraft.item.crafting.CraftingManager;

import java.util.*;

/**
 * Modular-Systems
 * Created by Paul Davis on 7/23/2015
 */
public class ContainerStorageCore extends BaseContainer {
    public TileStorageCore storage;
    protected List<Slot> displaySlots;
    protected List<Slot> allSlots;
    protected String currentSearch = "";

    protected InventoryCrafting craftingGrid;
    protected IInventory craftResult = new InventoryCraftResult();

    public ContainerStorageCore(InventoryPlayer playerInventory, IInventory ownerInventory, TileStorageCore tile) {
        super(playerInventory, ownerInventory);
        storage = tile;
        addInventoryGrid(25, 27, 11, 6);
        if(storage.getInventoryRowCount() > 6) {
            int slotId = inventorySlots.size();
            for(int i = 66; i < storage.getSizeInventory(); i++)
                addSlotToContainer(new RestrictedSlot(inventory, slotId++, -10000, -10000));
        }
        displaySlots = new ArrayList<>();
        displaySlots.addAll(inventorySlots);

        allSlots = new ArrayList<>();
        allSlots.addAll(displaySlots);

        addPlayerInventorySlots(storage.hasCraftingUpgrade() ? 8 : 44, 140);

        if(storage.hasCraftingUpgrade()) {
            craftingGrid = new DummyCraftingInventory(storage, this);
            addCraftingGrid(craftingGrid, 0, 180, 162, 3, 3);
            this.addSlotToContainer(new SlotCrafting(playerInventory.player, craftingGrid, craftResult, inventorySize + 1, 198, 140));
            onCraftMatrixChanged(craftingGrid);
        }
    }

    protected void addInventoryGrid(int xOffset, int yOffset, int width, int rows) {
        for (int y = 0, slotId = 0; y < rows; y++) {
            for (int x = 0; x < width; x++, slotId++) {
                addSlotToContainer(new RestrictedSlot(inventory, slotId,
                        xOffset + x * 18,
                        yOffset + y * 18));
            }
        }
    }

    public void scrollTo(float index) {
        int outsideDisplaySize = ((allSlots.size() / 11) + (allSlots.size() % 11 > 0 ? 1 : 0)) - 6;
        int currentTopRow = Math.round(index * (float)outsideDisplaySize);
        if(currentTopRow < 0)
            currentTopRow = 0;

        displaySlots.clear();

        for(int y = 0; y < 6; y++) {
            for(int x = 0; x < 11; x++) {
                if((x + ((y + currentTopRow) * 11)) >= allSlots.size())
                    break;
                Slot slot = this.allSlots.get((x + ((y + currentTopRow) * 11)));
                displaySlots.add(slot);
            }
        }

        updateDisplayedSlots();
    }

    public void keyTyped(String str) {
        currentSearch = str.toLowerCase();
        displaySlots.clear();
        displaySlots.addAll(allSlots);
        updateDisplayedSlots();
    }

    public void updateDisplayedSlots() {
        //Move it all off
        for (Slot displaySlot : allSlots) {
            displaySlot.xDisplayPosition = displaySlot.yDisplayPosition = -10000;
        }

        //Trim out what doesn't match the string
        for(Iterator<Slot> i = displaySlots.iterator(); i.hasNext();) {
            Slot slot = i.next();
            if(slot.getHasStack()) {
                if(currentSearch.startsWith("@")) {
                    GameRegistry.UniqueIdentifier id = GameRegistry.findUniqueIdentifierFor(slot.getStack().getItem());
                    if(currentSearch.length() > 1 && !id.modId.toLowerCase().contains(currentSearch.split("@")[1].toLowerCase()))
                        i.remove();
                }
                else if(!slot.getStack().getDisplayName().toLowerCase().contains(currentSearch.toLowerCase()))
                    i.remove();
            }
        }

        if(!currentSearch.isEmpty() && !currentSearch.equalsIgnoreCase(""))
            Collections.sort(displaySlots, new SlotComparator());

        for(int y = 0; y < 6; y++) {
            for(int x = 0; x < 11; x++) {
                if((x + (y * 11)) >= displaySlots.size())
                    break;
                Slot slot = this.displaySlots.get((x + (y * 11)));
                slot.xDisplayPosition = 25 + x * 18;
                slot.yDisplayPosition = 27 + y * 18;
            }
        }
    }

    public void detectAndSendChanges() {
        onCraftMatrixChanged(craftingGrid);
        super.detectAndSendChanges();
    }

    public void onCraftMatrixChanged(IInventory inv) {
        if(craftingGrid != null && inv == craftingGrid)
            craftResult.setInventorySlotContents(0, CraftingManager.getInstance().findMatchingRecipe(craftingGrid, storage.getWorldObj()));
    }

    public void addCraftingGrid(IInventory inventory, int startSlot, int x, int y, int width, int height) {
        int i = 0;
        for(int h = 0; h < height; h++) {
            for(int w = 0; w < width; w++) {
                this.addSlotToContainer(new Slot(inventory, startSlot + i++, x + (w * 18), y + (h * 18)));
            }
        }
    }

    public void clearCraftingGrid() {
        for(int i = inventorySlots.size() - 10; i < inventorySlots.size(); i++) {
            slotClick(i, 0, 1, ((InventoryPlayer)playerInventory).player);
        }
    }

    public boolean func_94530_a(ItemStack stack, Slot slot) {
        return slot.inventory != this.craftResult && super.func_94530_a(stack, slot);
    }

    @Override
    public ItemStack transferStackInSlot(EntityPlayer player, int slotId) {
        Slot slot = (Slot)inventorySlots.get(slotId);

        if (slot != null && slot.getHasStack()) {
            ItemStack itemToTransfer = slot.getStack();
            ItemStack copy = itemToTransfer.copy();
            if (slotId < inventorySize) {
                if (!mergeItemStackSafe(itemToTransfer, inventorySize, storage.hasCraftingUpgrade() ? inventorySlots.size() - 10 : inventorySlots.size(), true)) return null;
            }
            else if(slotId >= inventorySlots.size() - 10 && storage.hasCraftingUpgrade()) {
                if(slotId == inventorySlots.size() - 1) {
                    if(!mergeItemStackSafe(itemToTransfer, 0, storage.hasCraftingUpgrade() ? inventorySlots.size() - 10 : inventorySlots.size(), true)) return null;
                } else {
                    if(!mergeItemStackSafe(itemToTransfer, 0, storage.hasCraftingUpgrade() ? inventorySlots.size() - 10 : inventorySlots.size(), false)) return null;
                }
            }else if (!mergeItemStackSafe(itemToTransfer, 0, inventorySize, false)) return null;

            slot.onSlotChange(itemToTransfer, copy);
            slot.onPickupFromSlot(player, itemToTransfer);

            if (itemToTransfer.stackSize == 0) slot.putStack(null);
            else slot.onSlotChanged();
            onCraftMatrixChanged(craftingGrid);
            if (itemToTransfer.stackSize != copy.stackSize) return copy;
        }
        return null;

    }

    public static class DummyCraftingInventory extends InventoryCrafting {

        protected final TileStorageCore tile;
        protected final ContainerStorageCore container;
        protected ItemStack[] stacks;

        public DummyCraftingInventory(TileStorageCore tile, ContainerStorageCore container) {
            super(null, 3, 3);
            stacks = new ItemStack[9];
            this.tile = tile;
            this.container = container;
        }

        private void onCraftingChanged() {
            container.onCraftMatrixChanged(this);
        }

        @Override
        public int getSizeInventory() {
            return 9;
        }

        @Override
        public ItemStack getStackInSlot(int slot) {
            return slot >= getSizeInventory() ? null : tile.getCraftingGrid().getStackInSlot(slot);
        }

        @Override
        public ItemStack getStackInRowAndColumn(int row, int column) {
            if (row >= 0 && row < 3) {
                int k = row + column * 3;
                return getStackInSlot(k);
            } else
                return null;
        }

        @Override
        public ItemStack getStackInSlotOnClosing(int slot) {
            return tile.getCraftingGrid().getStackInSlot(slot);
        }

        @Override
        public ItemStack decrStackSize(int slot, int amount) {
            if(tile.getCraftingGrid().getStackInSlot(slot) != null) {
                ItemStack returnStack;
                if(tile.getCraftingGrid().getStackInSlot(slot).stackSize <= amount) {
                    returnStack = tile.getCraftingGrid().getStackInSlot(slot);
                    tile.getCraftingGrid().setStackInSlot(null, slot);
                    onCraftingChanged();
                    return returnStack;
                } else {
                    returnStack = tile.getCraftingGrid().getStackInSlot(slot).splitStack(amount);
                    if(tile.getCraftingGrid().getStackInSlot(slot).stackSize <= 0)
                        tile.getCraftingGrid().setStackInSlot(null, slot);
                    onCraftingChanged();
                    return returnStack;
                }
            }
            return null;
        }

        @Override
        public void setInventorySlotContents(int slot, ItemStack stack) {
            tile.getCraftingGrid().setStackInSlot(stack, slot);
            onCraftingChanged();
        }
    }

    protected class SlotComparator implements Comparator {
        @Override
        public int compare(Object o1, Object o2) {
            Slot thisSlot = (Slot)o1;
            Slot thatSlot = (Slot)o2;

            if(!thisSlot.getHasStack() && !thatSlot.getHasStack())
                return 0;
            else if(thisSlot.getHasStack() && !thatSlot.getHasStack())
                return -1;
            else if(!thisSlot.getHasStack())
                return 1;
            else if(thisSlot.getStack().getDisplayName().startsWith(currentSearch) && !thatSlot.getStack().getDisplayName().startsWith(currentSearch))
                return -1;
            else if(!thisSlot.getStack().getDisplayName().startsWith(currentSearch) && thatSlot.getStack().getDisplayName().startsWith(currentSearch))
                return 1;
            return 0;
        }
    }
}

File: src/main/scala/com/teambr/modularsystems/storage/container/ContainerStorageCore.scala
package com.teambr.modularsystems.storage.container

import java.util

import com.teambr.bookshelf.common.container.slots.IPhantomSlot
import com.teambr.bookshelf.util.InventoryUtils
import com.teambr.modularsystems.core.network.PacketManager
import com.teambr.modularsystems.storage.container.slot.{PlayerArmorSlot, SlotCraftingOutput, SlotStorageCore}
import com.teambr.modularsystems.storage.network.UpdateStorageContainer
import com.teambr.modularsystems.storage.tiles.TileStorageCore
import net.minecraft.entity.player.{EntityPlayer, InventoryPlayer}
import net.minecraft.inventory._
import net.minecraft.item.crafting.CraftingManager
import net.minecraft.item.{ItemFood, ItemStack, ItemTool}
import net.minecraft.nbt.NBTTagCompound
import net.minecraftforge.oredict.OreDictionary

import scala.collection.mutable.ArrayBuffer
import scala.util.control.Breaks._

/**
  * This file was created for Modular-Systems
  *
  * Modular-Systems is licensed under the
  * Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License:
  * http://creativecommons.org/licenses/by-nc-sa/4.0/
  *
  * @author Paul Davis "pauljoda"
  * @since 3/27/2016
  */
class ContainerStorageCore(val playerInventory: InventoryPlayer, val storageCore: TileStorageCore) extends Container {
    var INVENTORY_START = 0
    var INVENTORY_END = 0

    var CRAFTING_GRID_START = 0
    var CRAFTING_GRID_END = 0

    var PLAYER_INV_START_MAIN = 0
    var PLAYER_INV_END_MAIN = 0
    var PLAYER_INV_START_HOTBAR = 0
    var PLAYER_INV_END_HOTBAR = 0
    var PLAYER_INV_ARMOR_START = 0
    var PLAYER_INV_ARMOR_END = 0

    var rowStart = 0

    var filterString : String = ""
    lazy val listedItems = new java.util.ArrayList[ItemStack]()

    var isDirty = true
    var sendPacket = true

    addInventoryGrid(25, 27, 11, rowCount)

    lazy val craftMatrix = new InventoryCrafting(this, 3, 3)
    lazy val craftResult = new InventoryCraftResult

    if(storageCore.hasCraftingUpgrade)
        addCraftingGrid(98, 102, 3, 3)
    else
        CRAFTING_GRID_END = INVENTORY_END

    updateSlots()

    addPlayerInventorySlots(44, 160)

    if(storageCore.hasArmorUpgrade) {
        addPlayerArmor(15, 162)
    }

    def rowCount : Int = if(storageCore.hasCraftingUpgrade) 4 else 6

    /**
      * Adds an inventory grid to the container
      *
      * @param xOffset X pixel offset
      * @param yOffset Y pixel offset
      * @param width How many wide
      */
    def addInventoryGrid(xOffset : Int, yOffset : Int, width : Int, height : Int) : Unit = {
        INVENTORY_START = 0
        var slotId = -1
        for(y <- 0 until height) {
            for(x <- 0 until width) {
                slotId += 1
                addSlotToContainer(new SlotStorageCore(this, storageCore, slotId, xOffset + x * 18, yOffset + y * 18))
            }
        }
        INVENTORY_END = slotId
    }

    /**
      * Adds an inventory grid to the container
      *
      * @param xOffset X pixel offset
      * @param yOffset Y pixel offset
      * @param width How many wide
      */
    def addCraftingGrid(xOffset : Int, yOffset : Int, width : Int, height : Int) : Unit = {
        CRAFTING_GRID_START = INVENTORY_END + 1
        var slotId = -1
        for(y <- 0 until height) {
            for(x <- 0 until width) {
                slotId += 1
                addSlotToContainer(new Slot(this.craftMatrix, slotId, xOffset + x * 18, yOffset + y * 18))
            }
        }

        addSlotToContainer(new SlotCraftingOutput(playerInventory.asInstanceOf[InventoryPlayer].player,
            this.craftMatrix, this.craftResult, storageCore, 0, 171, 120))
        CRAFTING_GRID_END = CRAFTING_GRID_START + slotId + 1
    }

    /**
      * Adds the player offset with Y offset
      *
      * @param offsetY How far down
      */
    def addPlayerInventorySlots(offsetY: Int) {
        addPlayerInventorySlots(8, offsetY)
    }

    /**
      * Adds player inventory at location, includes space between normal and hotbar
      *
      * @param offsetX X offset
      * @param offsetY Y offset
      */
    def addPlayerInventorySlots(offsetX: Int, offsetY: Int): Unit = {
        PLAYER_INV_START_MAIN = CRAFTING_GRID_END + 1
        PLAYER_INV_END_MAIN = PLAYER_INV_START_MAIN
        for (row <- 0 until 3) {
            for (column <- 0 until 9) {
                addSlotToContainer(new Slot(playerInventory,
                    column + row * 9 + 9,
                    offsetX + column * 18,
                    offsetY + row * 18))
                PLAYER_INV_END_MAIN += 1
            }
        }

        PLAYER_INV_START_HOTBAR = PLAYER_INV_END_MAIN
        PLAYER_INV_END_HOTBAR = PLAYER_INV_START_HOTBAR
        for (slot <- 0 until 9) {
            addSlotToContainer(new Slot(playerInventory, slot, offsetX + slot * 18, offsetY + 58))
            PLAYER_INV_END_HOTBAR += 1
        }
    }

    def addPlayerArmor(offsetX: Int, offsetY: Int): Unit = {
        PLAYER_INV_ARMOR_START = PLAYER_INV_END_HOTBAR
        PLAYER_INV_ARMOR_END = PLAYER_INV_ARMOR_START
        for(x <- 0 until 4) {
            addSlotToContainer(new PlayerArmorSlot(playerInventory, 36 + x, offsetX, offsetY + (18 * (3 - x))))
            PLAYER_INV_ARMOR_END += 1
        }
    }

    override def canInteractWith(entityPlayer: EntityPlayer): Boolean = true
    /**
      * Looks for changes made in the container, sends them to every listener.
      */
    override def detectAndSendChanges() : Unit = {
        super.detectAndSendChanges()
        // Check us
        if(isDirty) {
            updateSlots()
            isDirty = false
            if(sendPacket) {
                PacketManager.net.sendToAll(new UpdateStorageContainer)
                sendPacket = false
            }
        }
        this.craftResult.setInventorySlotContents(0,
            CraftingManager.getInstance.findMatchingRecipe(this.craftMatrix, this.storageCore.getWorld))
    }

    override def func_184996_a(slotId : Int, dragType: Int, clickTypeIn: ClickType, player: EntityPlayer) : ItemStack = {
        isDirty = true
        if(!player.worldObj.isRemote)
            sendPacket = true
        if(slotId >= 0 && slotId < inventorySlots.size()) {
            val slot = inventorySlots.get(slotId)
            // Is a storage slot
            if(slot != null && slot.isInstanceOf[SlotStorageCore]) {
                // Wanting to quick move
                if(clickTypeIn == ClickType.QUICK_MOVE && slotId >= 0 && slotId < listedItems.size() && listedItems.get(slotId) != null) {
                    var stack = listedItems.get(slotId)
                    stack = storageCore.getStack(stack)
                    var i = 0
                    val list = storageCore.keysToList
                    breakable {
                        for (x <- 0 until list.size()) {
                            if (list.get(x) == stack) {
                                break
                            }
                            i += 1
                        }
                    }
                    stack = storageCore.getStack(stack)
                    val amount = storageCore.getInventory.get(stack)
                    stack = stack.copy()
                    stack.stackSize = amount
                    stack.stackSize = if(stack.stackSize >= stack.getMaxStackSize) stack.getMaxStackSize else stack.stackSize
                    if(stack != null && i < storageCore.getInventory.size()) {
                        val copy = stack.copy()
                        // Something was moved
                        if (mergeItemStackSafe(copy, PLAYER_INV_START_MAIN, PLAYER_INV_END_HOTBAR, reverse = true)) {
                            player.inventory.markDirty()
                            storageCore.extractItem(i, stack.stackSize - copy.stackSize, simulate = false)
                        }
                    }
                    return null
                }

                // Has something to drop
                if (player.inventory.getItemStack != null) {
                    val toInsert = player.inventory.getItemStack.copy()
                    if(dragType == 1)
                        toInsert.stackSize = 1
                    val size = toInsert.stackSize
                    val stack = storageCore.insertItem(-1, toInsert, simulate = false)
                    var toSet = player.inventory.getItemStack
                    toSet.stackSize -= (if(stack == null) size else size - stack.stackSize)
                    if(toSet.stackSize <= 0)
                        toSet = null
                    player.inventory.setItemStack(toSet)
                    return stack
                }

                // Trying to pick up something
                if (slotId >= 0 && slotId < listedItems.size() && listedItems.get(slotId) != null) {
                    var stack = listedItems.get(slotId)
                    stack = storageCore.getStack(stack)
                    var i = 0
                    val list = storageCore.keysToList
                    breakable {
                        for (x <- 0 until list.size()) {
                            if (list.get(x) == stack) {
                                break
                            }
                            i += 1
                        }
                    }
                    if(stack != null && i < storageCore.getInventory.size()) {
                        stack = storageCore.extractItem(i,
                            amount = if (dragType == 1)
                                if(stack.stackSize > 1)
                                    Math.ceil(Math.min(stack.stackSize, stack.getMaxStackSize) / 2).toInt
                                else
                                    stack.stackSize
                            else stack.getMaxStackSize, simulate = false)
                        player.inventory.setItemStack(stack)
                    }
                    return stack
                }
            }
            else if(slot != null && !slot.isInstanceOf[SlotStorageCore]) {
                val value =  super.func_184996_a(slotId, dragType, clickTypeIn, player)
                return value
            }
            null
        } else
            super.func_184996_a(slotId, dragType, clickTypeIn, player)
    }

    def updateSlots(): Unit = {
        listedItems.clear()
        storageCore.updateCachedSize()
        val list = storageCore.keysToList

        // Fill with all for filter
        val filteredList = new util.ArrayList[ItemStack]()
        for(x <- 0 until list.size())
            if(isValidForFilter(list.get(x)))
                filteredList.add(list.get(x))

        // Add what we can to display
        for(x <- 0 until PLAYER_INV_START_MAIN)
            if(filteredList.size() > x + (rowStart * 11))
                listedItems.add(filteredList.get(x + (rowStart * 11)))

        // Pad with nulls if need be
        for(i <- listedItems.size() until PLAYER_INV_START_MAIN)
            listedItems.add(null)
    }

    def isValidForFilter(stack : ItemStack, filter : String = filterString) : Boolean = {
        if(stack == null || stack.getItem == null || filterString.isEmpty)
            true
        else {
            if (filter.split(',').length == 1) {
                filter.charAt(0) match {
                    case '@' => // Matching by mod
                        if (filter.length > 1) {
                            // Check for food first, then tools
                            if (filter.substring(1).equalsIgnoreCase("food"))
                                return stack.getItem.isInstanceOf[ItemFood]
                            if (filter.substring(1).equalsIgnoreCase("tool"))
                                return stack.getItem.isInstanceOf[ItemTool]

                            // Check by mod
                            val uniqueName = stack.getItem.getRegistryName
                            val modName = uniqueName.split(':')(0)
                            modName.toLowerCase.contains(filter.substring(1).toLowerCase)
                        } else
                            true
                    case '$' => // Ore Dict
                        if (filter.length > 1) {
                            val oreDicts = OreDictionary.getOreIDs(stack)

                            // Go though all registered IDS
                            for (oreDictID <- oreDicts) {
                                val oreName = OreDictionary.getOreName(oreDictID)
                                if (oreName.toLowerCase.contains(filter.substring(1).toLowerCase))
                                    return true
                            }
                            false
                        }
                        else true
                    case '#' => // Stack size
                        if (filter.length > 2) {
                            filter.charAt(1) match {
                                case '>' =>
                                    if (filter.substring(2).matches("[0-9].*")) {
                                        val stackSize = Integer.valueOf(filter.substring(2))
                                        return stackSize < stack.stackSize
                                    }
                                case '<' =>
                                    if (filter.substring(2).matches("[0-9].*")) {
                                        val stackSize = Integer.valueOf(filter.substring(2))
                                        return stackSize > stack.stackSize
                                    }
                                case '=' =>
                                    if (filter.substring(2).matches("[0-9].*")) {
                                        val stackSize = Integer.valueOf(filter.substring(2))
                                        return stackSize == stack.stackSize
                                    }
                                case _ =>

                            }
                        }
                        true
                    case _ =>
                        stack.getDisplayName.toLowerCase.contains(filter.toLowerCase)
                }
            }
            else {
                for (string <- filterString.split(','))
                    if (!isValidForFilter(stack, string))
                        return false
                true
            }
        }
    }

    def scrollTo(index : Float) : Unit = {
        isDirty = true
        sendPacket = true
        val outsideDisplaySize = ((storageCore.getInventory.size() / 11) +
                (if(storageCore.getInventory.size() % 11 > 0)  1 else 0)) - rowCount
        this.rowStart = Math.round(index * outsideDisplaySize.toFloat)
        if(rowStart < 0)
            rowStart = 0
    }

    /**
      * Returns true if the player can "drag-spilt" items into this slot,. returns true by default. Called to check if
      * the slot can be added to a list of Slots to split the held ItemStack across.
      */
    override def canDragIntoSlot(slotIn: Slot) : Boolean = {
        !slotIn.isInstanceOf[SlotStorageCore]
    }

    protected def mergeItemStackSafe(stackToMerge: ItemStack, start: Int, stop: Int, reverse: Boolean): Boolean = {
        var inventoryChanged: Boolean = false
        val delta: Int = if (reverse) -1 else 1
        val slots: java.util.List[Slot] = inventorySlots
        if (stackToMerge.isStackable) {
            var slotId: Int = if (reverse) stop - 1 else start
            while (stackToMerge.stackSize > 0 && ((!reverse && slotId < stop) || (reverse && slotId >= start))) {
                val slot: Slot = slots.get(slotId)
                if (slot.isItemValid(stackToMerge) && !slot.isInstanceOf[IPhantomSlot]) {
                    val stackInSlot: ItemStack = slot.getStack
                    if (InventoryUtils.tryMergeStacks(stackToMerge, stackInSlot)) {
                        slot.onSlotChanged()
                        inventoryChanged = true
                    }
                }
                slotId += delta
            }
        }
        if (stackToMerge.stackSize > 0) {
            var slotId: Int = if (reverse) stop - 1 else start
            while ((!reverse && slotId < stop) || (reverse && slotId >= start)) {
                val slot: Slot = slots.get(slotId)
                if (slot.isItemValid(stackToMerge) && !slot.isInstanceOf[IPhantomSlot]) {
                    val stackInSlot: ItemStack = slot.getStack
                    if (stackInSlot == null) {
                        slot.putStack(stackToMerge.copy)
                        slot.onSlotChanged()
                        stackToMerge.stackSize = 0
                        return true
                    }
                }
                slotId += delta
            }
        }
        inventoryChanged
    }

    override def transferStackInSlot(player: EntityPlayer, slotId: Int): ItemStack = {
        isDirty = true
        if(!player.worldObj.isRemote)
            sendPacket = true
        val slot: Slot = inventorySlots.get(slotId)
        if (slot != null && slot.getHasStack && !slot.isInstanceOf[SlotStorageCore]) { // Not the inventory
        var itemToTransfer: ItemStack = slot.getStack
            val copy: ItemStack = itemToTransfer.copy

            // Attempt from outside into inventory
            val stackLeft = storageCore.insertItem(-1, itemToTransfer, simulate = false)
            // If a change happened, update the held stack
            if(stackLeft == null || stackLeft.stackSize != itemToTransfer.stackSize)
                itemToTransfer = stackLeft

            if(slot.isInstanceOf[SlotCraftingOutput])
                slot.onPickupFromSlot(playerInventory.asInstanceOf[InventoryPlayer].player, copy)

            if (itemToTransfer != null && slotId >= PLAYER_INV_START_HOTBAR && // From main inv into hotbar
                    !mergeItemStackSafe(itemToTransfer, PLAYER_INV_START_MAIN, PLAYER_INV_END_MAIN, reverse = false)) return null
            else if (itemToTransfer != null && slotId < PLAYER_INV_START_HOTBAR && slotId > INVENTORY_END // From hotbar into main inv
                    && !mergeItemStackSafe(itemToTransfer, PLAYER_INV_START_HOTBAR, PLAYER_INV_END_HOTBAR, reverse = false)) return null
            if (itemToTransfer == null || itemToTransfer.stackSize == 0) slot.putStack(null)
            else slot.onSlotChanged()
            if (itemToTransfer == null || itemToTransfer.stackSize != copy.stackSize) return copy
        }
        null
    }

    /**
      * Clears the crafting grid and returns true if gone
      *
      * @return
      */
    def clearCraftingGrid(packetNeeded : Boolean = true) : Boolean = {
        isDirty = true
        sendPacket = packetNeeded
        if(storageCore.hasCraftingUpgrade) {
            for (x <- CRAFTING_GRID_START to CRAFTING_GRID_END) {
                func_184996_a(x, 0, ClickType.QUICK_MOVE, playerInventory.asInstanceOf[InventoryPlayer].player)
            }
        }

        for(i <- 0 until this.craftMatrix.getSizeInventory)
            if(this.craftMatrix.getStackInSlot(i) != null)
                return false
        true
    }

    /**
      * Clears the crafting grid and returns true if gone
      *
      * @return
      */
    def clearPlayerInv(packetNeeded : Boolean = true) : Boolean = {
        isDirty = true
        sendPacket = packetNeeded

        for (x <- PLAYER_INV_START_MAIN until PLAYER_INV_END_MAIN)
            func_184996_a(x, 0, ClickType.QUICK_MOVE, playerInventory.asInstanceOf[InventoryPlayer].player)

        true
    }

    /**
      * Fills the crafting grid with the info sent by JEI
      *
      * @param recipe The recipe NBTTagCompound
      */
    def fillCraftingGrid(recipe : NBTTagCompound): Unit = {
        if(storageCore.hasCraftingUpgrade) {
            // Try and clear grid
            if(clearCraftingGrid()) { // Grid cleared continue
            val recipeBuffer = new ArrayBuffer[ArrayBuffer[ItemStack]]()
                // Create the slots objects
                for(x <- 0 until 9) {
                    // Get the list at the location
                    val list = recipe.getTagList("Stack:" + x, 10)

                    // If possible, read from the list
                    if(list != null && list.tagCount() > 0) {
                        // Create buffer that holds object
                        val buffer = new ArrayBuffer[ItemStack]()

                        // Iterate the list and add to held buffer
                        for(i <- 0 until list.tagCount())
                            buffer += ItemStack.loadItemStackFromNBT(list.getCompoundTagAt(i))

                        // Put slot stacks
                        recipeBuffer += buffer
                    } else // Nothing, put null to move around
                        recipeBuffer += new ArrayBuffer[ItemStack]()
                }

                // Try to fill
                for(x <- 0 until 9) {
                    val recipeStacks = recipeBuffer(x)
                    if(recipeStacks != null) {
                        for(stack <- recipeStacks) {
                            if (this.craftMatrix.getStackInSlot(x) == null && stack != null) {
                                // Get the key
                                var extractStack = storageCore.getStack(stack)

                                // If we have extracted something
                                if (extractStack != null) {
                                    var i = 0
                                    val list = storageCore.keysToList
                                    breakable {
                                        for (x <- 0 until list.size()) {
                                            if (list.get(x) == extractStack) {
                                                break
                                            }
                                            i += 1
                                        }
                                    }
                                    if (i < storageCore.getInventory.size()) {
                                        extractStack = storageCore.extractItem(i, stack.stackSize, simulate = false)
                                        this.craftMatrix.setInventorySlotContents(x, extractStack)
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
    }

    override def onContainerClosed(playerIn: EntityPlayer): Unit = {
        if(!playerIn.worldObj.isRemote) {
            var i = 0
            while (i < this.craftMatrix.getSizeInventory) {
                val stack = storageCore.insertItem(-1, this.craftMatrix.getStackInSlot(i), simulate = false)
                if (stack != null)
                    playerIn.dropPlayerItemWithRandomChoice(stack, false)
                i += 1
            }
        }
    }
}